src/pip/_internal/models/link.py

```diff
@@ -107,6 +95,20 @@ def is_hash_allowed(self, hashes: Optional[Hashes]) -> bool:
         return hashes.is_hash_allowed(self.name, hex_digest=self.value)
 
 
+@dataclass(frozen=True)
+class MetadataFile:
+    """Information about a core metadata file associated with a distribution."""
+
+    hashes: Optional[dict[str, str]]
+
+    # TODO: Do we care about stripping out unsupported hash methods?
```

Comment:
I don't have a strong preference since the values in ``_SUPPORTED_HASHES`` look fine to me (though it might be time to retire `md5` from that list). In theory people can use a number of other hashes that this method may not pick up, but I think that can/should be handled as a separate task of maintaining the list of supported hashes.

Reply:
Seems reasonable. This is mostly about handling garbage produced by incorrect servers (stuff like "sah256=abcbcba") than about limiting what hashes we use.

The attribute is already a weird tri-state thing (`None` = no data supplied, `MetadataFile(None)` = supplied with no hashes, and the full thing with hashes. There's also a dodgy `MetadataFile({})` which we get if we end up stripping everything, which is *probably* like `MetadataFile(None), but consumers might not like an empty dict of hashes.

I think what I should probably do is validate the input up front, rather than in this class. Then we just assert here that hashes must be `None` or a dictionary containing only one or more supported hash types.